### PR TITLE
Data Solution Extension (DSE) initial support

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds initial Data Solution Extension (DSE) support. It adds `vcd_dse_registry_configuration` resource and data source to handle registry configuration for packages and also `vcd_dse_solution_publish` for publishing available Data Solutions to tenants.

### CODE REVIEW

File: vcd/datasource_vcd_dse_registry_configuration.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			"rde_state": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Parent RDE state",
+			},
+		},
+	}
+}
+
+var datasourceDseContainerRegistry = &schema.Resource{
```
Comment (adambarreiro) [inline vcd/datasource_vcd_dse_registry_configuration.go:93]: Are these Required/Optional, or just Computed?

File: vcd/datasource_vcd_dse_registry_configuration.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			},
+			"container_registry": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: "Container registry configuration - only applicable for 'VCD Data Solutions'",
+				Elem:        datasourceDseContainerRegistry,
+			},
+			"compatible_version_constraints": {
+				Type:        schema.TypeSet,
+				Optional:    true,
```
Comment (adambarreiro) [inline vcd/datasource_vcd_dse_registry_configuration.go:73]: Should it be Computed?

File: vcd/datasource_vcd_dse_solution_publish.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+func datasourceVcdDsePublish() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: datasourceVcdDsePublishRead,
+
+		Schema: map[string]*schema.Schema{
+			"data_solution_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
```
Comment (adambarreiro) [inline vcd/datasource_vcd_dse_solution_publish.go:19]:
```suggestion
```

As there's no Create/Delete

File: vcd/datasource_vcd_dse_solution_publish.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+				Description: "The ID of Data Solution",
+			},
+			"org_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
```
Comment (adambarreiro) [inline vcd/datasource_vcd_dse_solution_publish.go:25]:
```suggestion
```
Same

File: vcd/resource_vcd_dse_registry_configuration.go [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+	// Data Solutions Operator (DSO) repository additionally has registry host configuration, it
+	// must also be removed
+	if dseEntryConfig.Name() == defaultDsoName {
+		cfg.Spec.DockerConfig = &types.DseDockerConfig{Auths: types.DseDockerAuths{}}
+	}
+
+	_, err = dseEntryConfig.Update(cfg)
+	if err != nil {
+		return diag.Errorf("error updating Data Solution R repository details for '%s': %s", d.Get("name").(string), err)
```
Comment (adambarreiro) [inline vcd/resource_vcd_dse_registry_configuration.go:331]:
```suggestion
		return diag.Errorf("error updating Data Solution repository details for '%s': %s", d.Get("name").(string), err)
```

File: website/docs/r/dse_registry_configuration.html.markdown [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+configuration, `package_repository` and `version` should be set.
+
+Helm based configurations apply to:
+
+* `MongoDB Community`
+* `Confluent Platform`
+* `MongoDB`
+
+
+Container bases configurations apply to:
```
Comment (adambarreiro) [inline website/docs/r/dse_registry_configuration.html.markdown:37]:
```suggestion
Container based configurations apply to:
```

**Review by adambarreiro — COMMENTED** [review]
First scan, still need to dig a bit more and run tests

File: vcd/datasource_vcd_dse_registry_configuration.go (hunk 1 above)
Comment (Didainius) [reply vcd/datasource_vcd_dse_registry_configuration.go:93]: This is a schema for data source - so yes - all computed.

File: vcd/datasource_vcd_dse_solution_publish.go (hunk 4 above)
Comment (Didainius) [reply vcd/datasource_vcd_dse_solution_publish.go:25]: Removed

File: vcd/resource_vcd_dse_registry_configuration.go (hunk 5 above)
Comment (Didainius) [reply vcd/resource_vcd_dse_registry_configuration.go:331]: Fixed.

File: website/docs/r/dse_registry_configuration.html.markdown (hunk 6 above)
Comment (Didainius) [reply website/docs/r/dse_registry_configuration.html.markdown:37]: Fixed.

File: examples/data-solution-extension/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# About
+
+This example is a 2 step approach to setup Data Solution Extension (DSE) from scratch and publish it
+to tenant for consumption. One **must** have Data Solution .iso file to use this example.
```
Comment (dataclouder) [inline examples/data-solution-extension/README.md:4]:
```suggestion
This example is a 2-step approach to setup Data Solution Extension (DSE) from scratch and publish it
to a tenant for consumption. One **must** have Data Solution .iso file to use this example.
```
Shall we provide the URL to the .ISO file?

File: examples/data-solution-extension/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+After executing both steps, then tenant should be able to login with newly provisioned user and see
```
Comment (dataclouder) [inline examples/data-solution-extension/README.md:6]:
```suggestion
After executing both steps, then the tenant should be able to login with newly provisioned user and see
```

File: examples/data-solution-extension/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+"Mongo DB Community" Data Solution visible (in UI it is at the top bar "More -> Data Solutions ->
+Solutions")
+
+**Note**. Technically, these steps could be combined into one, but a new connection is
+required after publishing Data Solution Instance to tenant.
```
Comment (dataclouder) [inline examples/data-solution-extension/README.md:11]:
```suggestion
required after publishing the Data Solution Instance to tenant.
```

File: examples/data-solution-extension/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+* Solution Landing Zone
+* Uploads Data Solution Extension (DSE) Add-On ISO image and configures it in Solution Landing Zone
+* Creates a Data Solutions Instance
+* Publishes Data Solutions Instance to tenant
+
+## Step 2
+
+Step 2 focuses on Data Solutions Add-On itself:
+
+* Create a global role that has combination of rights for Container Services Extension (CSE) and
```
Comment (dataclouder) [inline examples/data-solution-extension/README.md:25]:
```suggestion
* Create a global role that has a combination of rights for Container Services Extension (CSE) and
```

**Review by dataclouder — APPROVED** [review]

File: examples/data-solution-extension/README.md (hunk 7 above)
Comment (Didainius) [reply examples/data-solution-extension/README.md]: Fixed.

File: examples/data-solution-extension/README.md (hunk 8 above)
Comment (Didainius) [reply examples/data-solution-extension/README.md]: Fixed.

File: examples/data-solution-extension/README.md (hunk 9 above)
Comment (Didainius) [reply examples/data-solution-extension/README.md]: fixed

File: examples/data-solution-extension/README.md (hunk 10 above)
Comment (Didainius) [reply examples/data-solution-extension/README.md]: Fixed.

**Review by adambarreiro — APPROVED** [review]
Test passed and ran the examples, they work like a charm! 😃

File: website/docs/r/dse_registry_configuration.html.markdown [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  depends_on = [vcd_solution_add_on_instance_publish.public]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of Data Solution as it appears in repository configuration
+* `use_default_value` - (Optional) Defines if repository settings should be inherited from Data
```
Comment (lvirbalas) [inline website/docs/r/dse_registry_configuration.html.markdown:103]: Reading with fresh eyes this is obscure. I'm thinking what if it were named 'use_default_version' instead?

File: examples/data-solution-extension/step-2-publish-data-solution/main.tf [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+# Configure VCD Data Solutions (DSO) and Mongo DB Community version with default repositories
+resource "vcd_dse_registry_configuration" "dso" {
+  name              = "VCD Data Solutions"
+  use_default_value = true
+}
+
+resource "vcd_dse_registry_configuration" "mongodb-community" {
+  name              = "MongoDB Community"
+  use_default_value = true
```
Comment (lvirbalas) [inline examples/data-solution-extension/step-2-publish-data-solution/main.tf:59]: These caught my eye. Please see suggestion below in .markdown file.

File: website/docs/r/dse_registry_configuration.html.markdown (hunk 11 above)
Comment (Didainius) [reply website/docs/r/dse_registry_configuration.html.markdown:103]: It is not only the version, it is also repository address details (all the fields you can see as greyed out in these screenshots)

image
image

**Review by lvirbalas — APPROVED** [review]

**Pull request merged**